core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -0,0 +1,271 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import scala.Option;
+import scala.Tuple2;
+import scala.jdk.javaapi.CollectionConverters;
+import scala.runtime.BoxedUnit;
+
+/**
+ * A delayed share fetch operation has been introduced in case there is no share partition for which we can acquire records. We will try to wait
+ * for MaxWaitMs for records to be released else complete the share fetch request.
+ */
+public class DelayedShareFetch extends DelayedOperation {
+    private final SharePartitionManager.ShareFetchPartitionData shareFetchPartitionData;
+    private final ReplicaManager replicaManager;
+    private final Map<SharePartitionManager.SharePartitionKey, SharePartition> partitionCacheMap;
+
+    private static final Logger log = LoggerFactory.getLogger(DelayedShareFetch.class);
+
+    DelayedShareFetch(
+            SharePartitionManager.ShareFetchPartitionData shareFetchPartitionData,
+            ReplicaManager replicaManager,
+            Map<SharePartitionManager.SharePartitionKey, SharePartition> partitionCacheMap) {
+        super(shareFetchPartitionData.fetchParams().maxWaitMs, Option.empty());
+        this.shareFetchPartitionData = shareFetchPartitionData;
+        this.replicaManager = replicaManager;
+        this.partitionCacheMap = partitionCacheMap;
+    }
+
+    @Override
+    public void onExpiration() {
+    }
+
+    /**
+     * Complete the share fetch operation by fetching records for all partitions in the share fetch request irrespective
+     * of whether they have any acquired records. This is called when the fetch operation is forced to complete either
+     * because records can be acquired for some partitions or due to MaxWaitMs timeout.
+     */
+    @Override
+    public void onComplete() {
+        log.trace("onCompletion of delayed share fetch request for group {}, member {}, " +
+                        "topic partitions {}", shareFetchPartitionData.groupId(),
+                shareFetchPartitionData.memberId(), shareFetchPartitionData.partitionMaxBytes().keySet());
+
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = topicPartitionDataForAcquirablePartitions();
+        try {
+            if (topicPartitionData.isEmpty()) {
+                // No locks for share partitions could be acquired, so we complete the request with an empty response.
+                shareFetchPartitionData.future().complete(Collections.emptyMap());
+                return;
+            }
+            log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}",
+                    topicPartitionData, shareFetchPartitionData.groupId(), shareFetchPartitionData.fetchParams());
+
+            replicaManager.fetchMessages(
```

Comment:
When `onComplete` is called, we _must_ complete the purgatory entry. What happens if ReplicaManager#fetchMessages does not return synchronously here? I guess we would end up creating a fetch purgatory entry that is competed later on. 

If we are completing the purgatory entry due to a timeout, we shouldn't allow fetchMessages to create a new fetch purgatory entry as that would delay the response beyond the deadline. We should probably check the time remaining from MaxWaitMs when we call fetchMessages in `onComplete `

Reply:
We probably should just use r`eplicaManager.readFromLog` like `DelayedFetch`.

Reply:
Hi @mumrah @junrao , I have replaced `replicaManager.fetchMessages` with `replicaManager.readFromLog`